proposals/POEM-3-action-limit-for-namespace.md

```diff
@@ -0,0 +1,283 @@
+ │          │┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼│----------► │                        │
+ └──────────┴──────────────────────────────────┴────────────┴────────────────────────┘
+ ▲                                                          ▲                        ▲
+ │ system limit                                             │ namespace limit        │ system limit
+
+128M                                                      1024M                    2048M
+```
+
+## Proposed changes
+
+### 3 types of action limits
+
+There was only a system limit shared by all namespaces, but two more concepts for namespace limits are added.
+
+- (1) system limit: It can never be exceeded under any circumstances.
+- (2) namespace default limit: It can be used if a limit has not been set for a namespace.
+- (3) namespace limit: It can be set by a system administrator for a namespace and cannot exceed the range of the system limit.
+
+### Limit configs for namespace
+
+- The maxParameterSize, maxPayloadSize and truncationSize values are treated as `ByteSize` string. (e.g. 1 MB, 512 KB...)
+
+The following settings are new:
+
+config key             | Type                         | description
+---------------------- | ---------------------------- | ---------------
+minActionMemory        | integer (unit: MB)           | minimum action memory size for namespace
+maxActionMemory        | integer (unit: MB)           | maximum action memory size for namespace
+minActionLogs          | integer (unit: MB)           | minimum activation log size for namespace
+maxActionLogs          | integer (unit: MB)           | maximum activation log size for namespace
+minActionTimeout       | integer (unit: milliseconds) | minimum action time limit for namespace
+maxActionTimeout       | integer (unit: milliseconds) | maximum action time limit for namespace
+minActionConcurrency   | integer                      | minimum action concurrency limit for namespace
+maxActionConcurrency   | integer                      | maximum action concurrency limit for namespace
+maxParameterSize       | string  (format: ByteSize)   | maximum parameter size for namespace
+maxPayloadSize         | string  (format: ByteSize)   | maximum payload size for namespace
+truncationSize         | string  (format: ByteSize)   | activation truncation size for namespace
+
+
+### Limit document for CouchDB
+
+You can set namespace limits with `{namespace}/limits` document just like any other existing settings (invocationsPerMinute, concurrentInvocations..).
+
+```json
+{
+  "concurrentInvocations": 100,
+  "invocationsPerMinute": 100,
+  "firesPerMinute": 100,
+  "maxActionMemory": 1024,
+  "minActionMemory": 128,
+  "maxActionConcurrency": 400,
+  "minActionConcurrency": 1,
+  "maxActionLogs": 128,
+  "minActionLogs": 0,
+  "maxParameterSize": "1048576 B"
+}
+```
+
+#### Using wskadmin command (tool)
+- In general, it is recommended to use a wskadmin rather than modify the DB directly.
```

Comment:
I personally think it would be nice to have admin routes in the controller to update this data and you give just one namespace that the operators have access to authorization to hit the admin route path. `wskadmin` isn't really any different from editing the db directly security / authorization wise as it's just a cli tool that makes the db api calls.

That doesn't really preclude this proposal since this is just adding operator limits that can be configured that way, which is already done.

Reply:
I referred to the previous implementation for namespace limits, and I agree it is necessary to introduce a namespace with administrator privileges and an administrator API. I think it can be achieved through other proposals.

Reply:
Starting with `wskadmin` or modifications to couch is a fair transition. I also see the value of making this part of an OW Admin API. Regardless, I believe the `GET /api/v1/namespaces[_/limits]` should return not just the namespace name but the limits for that namespace as shown below.

Reply:
Using the wskadmin tool is the same as modifying the db directly, so I will delete the ambiguous phrase.